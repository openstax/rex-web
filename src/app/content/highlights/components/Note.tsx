import { HTMLTextAreaElement } from '@openstax/types/lib.dom';
import React from 'react';
import { useIntl } from 'react-intl';
import styled from 'styled-components/macro';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';
import { cardPadding, cardWidth } from '../constants';

interface Props {
  note: string;
  textareaRef: React.RefObject<HTMLTextAreaElement>;
  onChange: (note: string) => void;
  onFocus: () => void;
}

const noteMaxLength = 1000;

const width = cardWidth - cardPadding * 2;
// tslint:disable-next-line:variable-name
const TextArea = styled.textarea`
  display: block;
  min-height: 5.6rem;
  width: ${width}rem;
  max-height: 30rem;
  max-width: ${width}rem;
  min-width: ${width}rem;
  border: 1px solid ${theme.color.neutral.formBorder};
  padding: ${cardPadding}rem;
  ${textStyle}
  color: ${theme.color.text.label};
  font-size: 1.4rem;
  font-family: inherit;
  line-height: 2rem;
  font-weight: normal;
`;

// tslint:disable-next-line:variable-name
const Note = ({onChange, onFocus, note, textareaRef}: Props) => {
  const setTextAreaHeight = React.useCallback(() => {
    const element = textareaRef.current;
    if (!element) {
      return;
    }

    if (element.scrollHeight > element.offsetHeight) {
      element.style.height = `${element.scrollHeight + 5}px`;
    }
  }, [textareaRef]);

  React.useEffect(setTextAreaHeight, [note, setTextAreaHeight]);

  return <TextArea
    ref={textareaRef}
    value={note}
    onFocus={onFocus}
    maxLength={noteMaxLength}
    onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => {
      onChange(e.target.value);
    }}
    placeholder={useIntl().formatMessage({id: 'i18n:highlighting:card:placeholder'})}
  />;
};

export default Note;
