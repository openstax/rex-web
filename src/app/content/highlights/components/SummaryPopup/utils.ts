import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import React from 'react';
import { useSelector } from 'react-redux';
import { systemQueryParameters } from '../../../../navigation/selectors';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const useCreateHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const [link, setLink] = React.useState('');
  const systemQueryParams = useSelector(systemQueryParameters);

  React.useEffect(() => {
    const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
    const data: Record<string, string> = {
      ...systemQueryParams,
      target: JSON.stringify({ id, type: 'highlight' }),
    };
    const target = queryString.stringify(data) + `#${elementId}`;
    setLink(`${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`);
  }, [id, book, elementId, sourceId, systemQueryParams]);

  return link;
};
