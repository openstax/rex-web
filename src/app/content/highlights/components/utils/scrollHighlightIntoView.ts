import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import { scrollIntoView } from '../../../../domUtils';

export default (highlight: Highlight, cardRef?: React.RefObject<HTMLElement>) => {
  const firstElement = highlight.elements[0] as HTMLElement;
  const lastElement = highlight.elements[highlight.elements.length - 1] as HTMLElement;
  const otherElements = [lastElement];
  if (cardRef?.current) {
    otherElements.push(cardRef.current);
  }
  scrollIntoView(firstElement, otherElements);
};
