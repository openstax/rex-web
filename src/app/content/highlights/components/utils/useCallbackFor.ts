import React from 'react';

// Some functions get recreated but shouldn't be reacted to
// This makes a non-changing callback
export default function useCallbackFor<T extends Array<unknown>>(fn: (...args: T) => unknown) {
    const ref = React.useRef(fn);
    const cb = React.useCallback(
        (...args: T) => ref.current(...args),
        []
    );

    ref.current = fn;
    return cb;
}
