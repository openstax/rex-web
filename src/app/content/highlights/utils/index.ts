export {
  getHighlightLocationFilters,
  getHighlightColorFiltersWithContent,
  getHighlightLocationFilterForPage,
  getHighlightLocationFiltersWithContent,
  sectionIsHighlightLocationFilter
} from './locationFiltersUtils';

export {
  addToTotalCounts,
  updateInTotalCounts,
  removeFromTotalCounts,
  addSummaryHighlight,
  getHighlightByIdFromSummaryHighlights,
  removeSummaryHighlight,
  updateSummaryHighlight,
  updateSummaryHighlightsDependOnFilters,
  getSortedSummaryHighlights
} from './summaryHighlightsUtils';
