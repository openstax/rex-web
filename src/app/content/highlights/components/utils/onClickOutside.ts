import { Element, EventListenerOptions, HTMLElement, MouseEvent } from '@openstax/types/lib.dom';
import curry from 'lodash/fp/curry';
import React from 'react';
import { isHtmlElement } from '../../../../guards';
import { assertDocument, assertWindow } from '../../../../utils';

const targetBelongsToElement = curry((event: MouseEvent, el: React.RefObject<HTMLElement> | HTMLElement): boolean =>
  isHtmlElement(el)
    ? el.contains(event.target as Element)
    : el.current
      ? el.current.contains(event.target as Element)
      : false
);

const onClickOutside = (
  element: React.RefObject<HTMLElement> | Array<React.RefObject<HTMLElement> | HTMLElement>,
  isEnabled: boolean,
  cb: (ev: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => () => {
  if (typeof document === 'undefined') {
    return () => null;
  }

  const ifOutside = (e: MouseEvent) => {
    if (!(e.target instanceof assertWindow().Element)) {
      return;
    } else if (Array.isArray(element) && element.some(targetBelongsToElement(e))) {
      return;
    } else if (!Array.isArray(element) && targetBelongsToElement(e, element)) {
      return;
    }
    cb(e);
  };

  if (isEnabled) {
    document.addEventListener('click', ifOutside, eventOptions);
  }

  return () => assertDocument().removeEventListener('click', ifOutside, eventOptions);
};

export const useOnClickOutside = (
  element: React.RefObject<HTMLElement> | Array<React.RefObject<HTMLElement> | HTMLElement>,
  isEnabled: boolean,
  cb: (e: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  React.useEffect(onClickOutside(element, isEnabled, cb, eventOptions), [element, cb, isEnabled]);
};

export default onClickOutside;

const isRefWithHtmlElement = (el: any): el is React.RefObject<HTMLElement> => {
  return el instanceof Object && isHtmlElement(el.current);
};

export const isElementForOnClickOutside = (el: any): el is HTMLElement | React.RefObject<HTMLElement> => {
  return isHtmlElement(el) || isRefWithHtmlElement(el);
};
