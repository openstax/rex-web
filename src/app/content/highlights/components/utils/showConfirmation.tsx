import React from 'react';
import ReactDOM from 'react-dom';
import { RawIntlProvider } from 'react-intl';
import uuid from 'uuid/v4';
import { AppServices } from '../../../../types';
import { assertDocument, assertNotNull } from '../../../../utils';
import ConfirmationModal from '../ConfirmationModal';

export default async(services: AppServices) => {
  const document = assertDocument();
  const domNode = document.createElement('div');

  domNode.id = `dialog-${uuid()}`;
  const root = assertNotNull(document.getElementById('root'), 'root element not found');
  root.insertAdjacentElement('afterend', domNode);

  const userChoice: boolean = await new Promise((resolve) => {
    const confirm = () => resolve(true);
    const deny = () => resolve(false);
    ReactDOM.render(
      <RawIntlProvider value={services.intl}>
        <ConfirmationModal deny={deny} confirm={confirm} />
      </RawIntlProvider>,
      domNode
    );
  });

  ReactDOM.unmountComponentAtNode(domNode);
  assertNotNull(domNode.parentElement, 'parent element not found').removeChild(domNode);

  return userChoice;
};
