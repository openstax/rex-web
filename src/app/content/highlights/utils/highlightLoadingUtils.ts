import {
  GetHighlightsRequest,
  GetHighlightsSourceTypeEnum,
  Highlights,
} from '@openstax/highlighter/dist/api';
import omit from 'lodash/fp/omit';
import { AppServices } from '../../../types';
import { assertDefined } from '../../../utils';
import { LocationFilters } from '../../components/popUp/types';
import { maxHighlightsApiPageSize } from '../../constants';
import { Book } from '../../types';
import { stripIdVersion } from '../../utils/idUtils';
import { CountsPerSource, HighlightData, SummaryHighlightsPagination } from '../types';
import { addSummaryHighlight, getHighlightLocationFilterForPage } from './';
import { getNextPageSources, incrementPage } from './paginationUtils';

const extractDataFromHighlightClientResponse = (highlightsResponse: Highlights) => {
  // TODO - change swagger so none of this is nullable
  const data = assertDefined(highlightsResponse.data, 'response from highlights api is invalid') as HighlightData[];
  const meta = assertDefined(highlightsResponse.meta, 'response from highlights api is invalid');
  const page = assertDefined(meta.page, 'response from highlights api is invalid');
  const perPage = assertDefined(meta.perPage, 'response from highlights api is invalid');
  const totalCount = assertDefined(meta.totalCount, 'response from highlights api is invalid');

  return {
    data,
    meta,
    page,
    perPage,
    totalCount,
  };
};

const getNewSources = (
  book: Book | undefined,
  omitSources: string[],
  countsPerSource: CountsPerSource,
  pageSize?: number
) => {
  const remainingCounts = omit(omitSources, countsPerSource);
  return book ? getNextPageSources(remainingCounts, book.tree, pageSize) : [];
};

export const formatReceivedHighlights = (
  highlights: HighlightData[],
  locationFilters: LocationFilters
) => highlights.reduce((result, highlight) => {
  const pageId = stripIdVersion(highlight.sourceId);
  const location = assertDefined(
    getHighlightLocationFilterForPage(locationFilters, pageId),
    'highlight is not in a valid location'
  );
  const locationFilterId = stripIdVersion(location.id);

  return addSummaryHighlight(result, {
    highlight,
    locationFilterId,
    pageId,
  });
}, {} as ReturnType<typeof addSummaryHighlight>);

const fetchHighlightsForSource = async({
  highlightClient,
  prevHighlights,
  book,
  colors,
  sets,
  pagination,
}: {
  highlightClient: AppServices['highlightClient'],
  prevHighlights?: HighlightData[],
  book: Book,
  pagination: NonNullable<SummaryHighlightsPagination>,
  colors: NonNullable<GetHighlightsRequest['colors']>;
  sets?: GetHighlightsRequest['sets'];
}) => {
  const highlightsResponse = await highlightClient.getHighlights({
    colors,
    scopeId: book.id,
    sets,
    sourceType: GetHighlightsSourceTypeEnum.OpenstaxPage,
    ...pagination,
  });

  const {data, perPage, totalCount} = extractDataFromHighlightClientResponse(highlightsResponse);

  const loadedResults = (pagination.page - 1) * perPage + data.length;
  const nextPagination = loadedResults < totalCount
    ? pagination
    : null;

  return {
    highlights: prevHighlights ? [...prevHighlights, ...data] : data,
    pagination: nextPagination,
  };
};

export const loadUntilPageSize = async({
  previousPagination,
  ...args
}: {
  previousPagination: SummaryHighlightsPagination,
  book: Book | undefined,
  highlightClient: AppServices['highlightClient'],
  colors: NonNullable<GetHighlightsRequest['colors']>;
  sets?: GetHighlightsRequest['sets'];
  sourcesFetched: string[],
  countsPerSource: CountsPerSource,
  highlights?: HighlightData[]
  pageSize?: number,
}): Promise<{pagination: SummaryHighlightsPagination, highlights: HighlightData[]}> => {
  const {page, sourceIds, perPage} = previousPagination
    ? incrementPage(previousPagination)
    : {
      page: 1,
      perPage: args.pageSize || maxHighlightsApiPageSize,
      sourceIds: getNewSources(args.book, args.sourcesFetched, args.countsPerSource, args.pageSize),
    };

  if (!args.book || sourceIds.length === 0) {
    return {pagination: null, highlights: args.highlights || []};
  }

  const {highlights, pagination} = await fetchHighlightsForSource({
    book: args.book,
    colors: args.colors,
    highlightClient: args.highlightClient,
    pagination: {page, sourceIds, perPage},
    prevHighlights: args.highlights,
    sets: args.sets,
  });

  if (highlights.length < perPage || !args.pageSize) {
    return loadUntilPageSize({
      ...args,
      highlights,
      previousPagination: pagination,
      sourcesFetched: [...args.sourcesFetched, ...sourceIds],
    });
  }
  return {pagination, highlights};
};

export const loadAllHighlights = async({
  highlightClient,
  book,
  pagination,
  sets,
}: {
  highlightClient: AppServices['highlightClient'];
  book: Book;
  pagination: NonNullable<SummaryHighlightsPagination>;
  sets?: GetHighlightsRequest['sets']
}): Promise<HighlightData[]> => {
  const highlightsResponse = await highlightClient.getHighlights({
    ...pagination,
    perPage: maxHighlightsApiPageSize,
    scopeId: book.id,
    sets,
    sourceType: GetHighlightsSourceTypeEnum.OpenstaxPage,
  });

  const {data, page, perPage, totalCount} = extractDataFromHighlightClientResponse(
    highlightsResponse);
  const loadedResults = (page - 1) * perPage + data.length;

  if (loadedResults < totalCount) {
    const moreResults = await loadAllHighlights({
      book,
      highlightClient,
      pagination: incrementPage(pagination),
      sets,
    });
    return [...data, ...moreResults];
  } else {
    return data;
  }
};
