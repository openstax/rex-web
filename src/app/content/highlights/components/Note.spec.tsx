import React from 'react';
import renderer from 'react-test-renderer';
import { renderToDom } from '../../../../test/reactutils';
import TestContainer from '../../../../test/TestContainer';
import { assertDocument } from '../../../utils/browser-assertions';
import Note from './Note';

describe('Note', () => {
  it('matches snapshot', () => {
    const textarea = assertDocument().createElement('textarea');

    const component = renderer.create(<TestContainer>
      <Note textareaRef={{ current: textarea }} note='' onChange={() => null} onFocus={() => null} />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls onChange', () => {
    const textareaElement = assertDocument().createElement('textarea');

    const onChange = jest.fn();
    const component = renderer.create(<TestContainer>
      <Note textareaRef={{ current: textareaElement }} note='' onChange={onChange} onFocus={() => null} />
    </TestContainer>);

    const textarea = component.root.findByType('textarea');

    renderer.act(() => {
      textarea.props.onChange({
        target: {
          value: 'asdf',
        },
      });
    });

    expect(onChange).toHaveBeenCalledWith('asdf');
  });

  it('resizes on update when necessary', () => {
    const textarea = assertDocument().createElement('textarea');

    const {node, root} = renderToDom(<TestContainer>
      <Note textareaRef={{ current: textarea }} note='' onChange={() => null} onFocus={() => null} />
    </TestContainer>);

    Object.defineProperty(node, 'scrollHeight', { value: 100 });
    Object.defineProperty(node, 'offsetHeight', { value: 50 });

    renderToDom(<TestContainer>
      <Note textareaRef={{ current: textarea }} note='asdf' onChange={() => null} onFocus={() => null} />
    </TestContainer>, root);

    expect(node.style.height).toEqual('105px');
  });

  it('doesn\'t resize on update when unneccessary', () => {
    const textarea = assertDocument().createElement('textarea');

    const {node, root} = renderToDom(<TestContainer>
      <Note textareaRef={{ current: textarea }} note='' onChange={() => null} onFocus={() => null} />
    </TestContainer>);

    Object.defineProperty(node, 'scrollHeight', { value: 50 });
    Object.defineProperty(node, 'offsetHeight', { value: 50 });

    renderToDom(<TestContainer>
      <Note textareaRef={{ current: textarea }} note='asdf' onChange={() => null} onFocus={() => null} />
    </TestContainer>, root);

    expect(node.style.height).toEqual('');
  });
});
