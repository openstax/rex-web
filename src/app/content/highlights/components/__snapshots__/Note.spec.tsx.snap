// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note matches snapshot 1`] = `
.c0 {
  display: block;
  min-height: 5.6rem;
  width: 18.4rem;
  max-height: 30rem;
  max-width: 18.4rem;
  min-width: 18.4rem;
  border: 1px solid #d5d5d5;
  padding: 0.8rem;
  color: #424242;
  color: #6f6f6f;
  font-size: 1.4rem;
  font-family: inherit;
  line-height: 2rem;
  font-weight: normal;
}

<textarea
  className="c0"
  maxLength={1000}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder="Add a note..."
  value=""
/>
`;
